docs/design/features/bridging-constraints.md

```diff
@@ -0,0 +1,110 @@
+
+It would greatly simplify the codegen impact of this if we were to limit the amount of places within a method body that are allowed to break the constraints. The least impactful and potentially sufficiently powerful change would be allowing to break constraints on `call` instruction.
+
+The proposal is to:
+
+* Stop doing the verification-level constraint validation when compiling `call` (i.e. the check on uninstantiated forms).
+* Keep constraint validation of the instantiated form.
+* If the instantiated form of the call target doesn't satisfy constraint, replace the call with a call to a throw helper (we can likely reuse infrastructure around `CORINFO_ACCESS_ILLEGAL`).
+* If the constraint validation requires a runtime check due to shared code, delay validation to run time (this can potentially be optimized out if the potential constraint violation is already in a guarded block).
+
+### Alternative option: allow loading constraint-failing MethodTables
+
+Limiting the constraint-violating code to `call` instruction is required to simplify code generation. Allowing `MethodTable`s that don't satisfy their own constraints to load sounds like trouble. We could potentially allow them to load and then try to prevent them from leaking out. Calling this out for completeness, I don't know if it's a good idea.
+
+### Potential shortcut: Stop validating constraints
+
+One more avenue could be to drop all constraint checks within the runtime and make constraints source compiler's concern only. We already have a constraint (`unmanaged` in C#) that is only known to the source compiler and not enforced at run time. This would likely require some thoughts around failure modes when invalid IL within a method body IL is encountered. This document is not going to explore this direction further since the assumption is that we do not want to drop constraint checks from the runtime, we just want to delay them.
+
+## API proposal
+
+Runtime support for this will be indicated by following property:
+
+```diff
+namespace System.Runtime.CompilerServices
+{
+    public static partial class RuntimeFeature
+    {
++        /// <summary>
++        /// Represents a runtime feature where constraint validation can be delayed until run time.
++        /// </summary>
++        public const string DelayedConstraintCheck = nameof(DelayedConstraintCheck);
+    }
+}
+```
+
+C# compiler will also need APIs to check for constraints at runtime. Some constraints could be checked using existing reflection APIs; it's questionable if we're okay with the perf characteristics or whether we want new APIs.
+
+* `new()` constraint: add `bool RuntimeHelpers.SatisfiesNewConstraint<T>()` (no existing reflection API for this, we need to check if it's a valuetype or a non-abstract type with public parameterless constructor)
+* `struct` constraint: add `bool RuntimeHelpers.IsValueType<T>()`, or reuse `typeof(T).IsValueType`
```

Comment:
We probably need a special API with different name because ```Nullable<T>``` is a value type, but it doesn't satisfy `struct` constraint.